grpc_unary_attestation/src/client.rs

```diff
@@ -0,0 +1,118 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::proto::{unary_session_client::UnarySessionClient, UnaryRequest};
+use anyhow::Context;
+use oak_functions_abi;
+use oak_remote_attestation::handshaker::{
+    AttestationBehavior, ClientHandshaker, Encryptor, ServerIdentityVerifier,
+};
+use tonic::transport::Channel;
+
+type SessionId = u64;
+
+/// gRPC Attestation Service client implementation.
+pub struct AttestationClient {
+    session_id: SessionId,
+    encryptor: Encryptor,
+    client: UnarySessionClient<Channel>,
+}
+
+impl AttestationClient {
+    pub async fn create(
+        uri: &str,
+        expected_tee_measurement: &[u8],
+        server_verifier: ServerIdentityVerifier,
+    ) -> anyhow::Result<Self> {
+        let session_id = 0;
+        let channel = Channel::from_shared(uri.to_string())
+            .context("Couldn't create gRPC channel")?
+            .connect()
+            .await?;
+        let mut client = UnarySessionClient::new(channel);
+
+        let mut handshaker = ClientHandshaker::new(
+            AttestationBehavior::create_peer_attestation(expected_tee_measurement),
+            server_verifier,
+        );
+        let client_hello = handshaker
+            .create_client_hello()
+            .context("Couldn't create client hello message")?;
+
+        let mut response = client
+            .message(UnaryRequest {
+                body: client_hello,
+                session_id,
+            })
+            .await
+            .context("Couldn't send client hello message")?
+            .into_inner();
+
+        while !handshaker.is_completed() {
+            let request = handshaker
+                .next_step(&response.body)
+                .context("Couldn't process handshake message")?;
+
+            if let Some(request) = request {
+                response = client
+                    .message(UnaryRequest {
+                        body: request,
+                        session_id,
+                    })
+                    .await
+                    .context("Couldn't send client hello message")?
+                    .into_inner();
+            }
+        }
```

Comment:
This loop matches the streaming client pretty closely, but I'm not 100% happy with it. The reason is it expects an implicit correlation between `handshaker.is_completed()` being false (the loop condition), and the handshaker returning `Some` request from `handshaker.next_step(&response.body)`. If that's ever not the case there will be an infinite loop. Tracked this issue for later refactoring in #2608.